README.md
# Hanginglist

A Flutter library to create a `HangingList` with animation on scrolling. You can also create `HangingItems` which you can rotate like a hanger and play with their flip property.
The idea is taken from hanging clothes in almirah.

![Hanging List Demo](demo/hangingList.gif) ![Flip Item Demo](demo/flipAnimation.gif)

[![Hanging List Demo Video](https://i9.ytimg.com/vi/eTNpH7j6xMU/mq3.jpg?sqp=CLCp1e8F&rs=AOn4CLDP1CiygUqyLlzz8In2Bmd0012fSw)](https://youtu.be/kzgtRHAu7eg)

## Getting Started

#### 1. Add dependency to your `pubspec.yaml`

```yaml
dependencies:
  hanginglist: ^0.1.1
```

#### 2. Import it

```dart
import 'package:hanginglist/hanginglist.dart'
```

#### 3. Use it. See the examples folder for an ... example.

## Parameters

```dart
    @required this.list; //list of items
    @required this.frontSide; //widget to show on front side of card
    @required this.backSide;  //widget to show on back side of card
    @required this.moveList;  // bool value whether to move whole list with animation
    @required this.moveListItem;  // bool value whether to move list item with animation
    @required this.height; //height of card
    @required this.width;  //width of card
```

`HangingList<T>` is a widget which builds list of cards.

```dart
HangingList({
    @required this.list,
    @required this.frontSide,
    @required this.backSide,
    @required this.moveList,
    @required this.moveListItem,
    @required this.height,
    @required this.width,
  });
```

## Example

```dart
//Initialise the HangingList widget where you want to use it.
HangingList<HangingObject>(
        list: list,
        frontSide: frontSide,
        backSide: backSide,
        moveList: true,
        moveListItem: true,
        height: screenHeight / 2,
        width: screenWidth / 1.3,
      ),

// create a generic function frontSide which excepts the type of list you want in <Object> parameters and return the widget you want to show in the card.

Widget frontSide<HangingObject>(obj) {
    return yourWidget();
}
// Similarly create backSide which will show when you flip the card.
Widget backSide<HangingObject>(obj) {
    return yourWidget();
}
// Also do not forget to initialise the list with type params.
List<HangingObject> list = [
    HangingObject(
        'Chicken',
        'McDonald`s',
        ['Chicken McNuggets', 'Big Mac', 'Fries'],
        'Fusion Mall,2ndStage, Kormangala',),
    HangingObject(
        'Burger',
        'Taco Bell',
        ['Doritos Locos Tacos', 'Freezes', 'Nacho Fries', 'The Meximelt'],
        'Ashwini Complex, Indranagar',)];
```

This package is currently in development. To enjoy all the features you can edit this package.

## FAQ

**Q:** How can I change PageView Builder to ListViewBuilder ?

**A:** Go to HangingList widget and chage PageView builder to list view builder and change PageController to ScrollController.

##

**Q:** How can I add more gestures to List Items ?

**A:** Go to HangingList widget > HangingItem widget and in gestureDetector add your own gesture functionalities.

##

**Q:** How can I change List Item decoration ?

**A:** Go to the HangingItem widget > Animation Card widget > HangTransform widget and add your decoration there.


CHANGELOG.md
## [0.1.1]: Dec 19, 2019.

- PageView builder flickering bug fixed.


test/hanginglist_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:hanginglist/hanginglist.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
  });
}


lib/hanginglist.dart
library hanginglist;

import 'package:flutter/material.dart';
import 'package:flutter/rendering.dart';
import 'dart:math';

import 'package:hanginglist/widgets/hangingItem.dart';

class HangingList<T> extends StatefulWidget {
  final List list;
  final Widget Function<T>(T) frontSide;
  final Widget Function<T>(T) backSide;
  final bool moveList, moveListItem;
  final double height, width;

  HangingList({
    @required this.list,
    @required this.frontSide,
    @required this.backSide,
    @required this.moveList,
    @required this.moveListItem,
    @required this.height,
    @required this.width,
  });
  @override
  _HangingListState createState() => _HangingListState();
}

class _HangingListState extends State<HangingList>
    with SingleTickerProviderStateMixin {
  double _angle = 0;
  AnimationController animationController;
  Animation<double> animation;

  PageController _pageController;
  bool endAnim = false;

  @override
  void initState() {
    super.initState();
    _pageController = PageController()
      ..addListener(() {
        setState(() {});
      });

    animationController =
        AnimationController(duration: Duration(milliseconds: 1500), vsync: this)
          ..addListener(() {});
  }

  initialiseAngle(bool left) {
    double angle = pi / 16;
    if (left == true) {
      angle = -angle;
    }
    animation = Tween(begin: angle, end: 0.0).animate(CurvedAnimation(
      parent: animationController,
      curve: Curves.bounceOut,
    ))
      ..addListener(() {
        setState(() {});
      });
  }

  @override
  Widget build(BuildContext context) {
    return NotificationListener<ScrollNotification>(
      onNotification: widget.moveList
          ? (notification) {
              if (notification is ScrollStartNotification) {
              } else if (notification is ScrollUpdateNotification) {
                if (_pageController.position.userScrollDirection ==
                    ScrollDirection.forward) {
                  _angle = pi / 16;
                } else {
                  _angle = -pi / 16;
                }
              } else if (notification is ScrollEndNotification) {
                try {
                  endAnim = true;

                  if (_pageController.position.userScrollDirection ==
                      ScrollDirection.forward) {
                    initialiseAngle(true);
                  } else {
                    initialiseAngle(false);
                  }
                  animationController
                      .forward(from: 0.0)
                      .orCancel
                      .then((onValue) {
                    _angle = 0;
                    endAnim = false;
                  });
                } on TickerCanceled {}
              }
              return false;
            }
          : null,
      child: PageView.builder(
          scrollDirection: Axis.horizontal,
          itemCount: widget.list.length,
          controller: widget.moveList ? _pageController : null,
          itemBuilder: (context, index) {
            if (endAnim) {
              _angle = animation.value;
            } else {
              _angle = _angle;
            }
            return Transform.rotate(
                angle: widget.moveList ? _angle : 0,
                child: Container(
                    padding: EdgeInsets.all(10),
                    child: HangingItem(
                        widget.frontSide(widget.list[index]),
                        widget.backSide(widget.list[index]),
                        widget.moveListItem,
                        widget.height,
                        widget.width)));
          }),
    );
    // )
  }

  @override
  void dispose() {
    animationController.dispose();
    super.dispose();
  }
}


lib/widgets/hangingItem.dart
import 'dart:math';
import 'package:flutter/material.dart';

import 'animationCard.dart';

class HangingItem extends StatefulWidget {
  final Widget front, back;
  final bool moveListItem;
  final double _height, _width;

  HangingItem(
      this.front, this.back, this.moveListItem, this._height, this._width);
  @override
  _HangingItemState createState() => _HangingItemState();
}

typedef void BoolCallback(bool isFront);

class _HangingItemState extends State<HangingItem>
    with TickerProviderStateMixin {
  AnimationController animationController;
  AnimationController _flipController;

  Animation<double> _frontScale;
  Animation<double> _backScale;

  Animation<double> animation;

  double _angle = 0;
  Offset _start, _end;
  bool startAnim = false, isFront = true;

  VoidCallback onFlip;
  BoolCallback onFlipDone;
  bool flipOnTouch;

  @override
  void initState() {
    super.initState();

    animationController =
        AnimationController(duration: Duration(milliseconds: 1000), vsync: this)
          ..addListener(() {});
    _flipController = new AnimationController(
      vsync: this,
      duration: const Duration(milliseconds: 800),
    );

    _frontScale = TweenSequence(
      <TweenSequenceItem<double>>[
        TweenSequenceItem<double>(
          tween: Tween(begin: 0.0, end: pi / 2)
              .chain(CurveTween(curve: Curves.easeIn)),
          weight: 50.0,
        ),
        TweenSequenceItem<double>(
          tween: ConstantTween<double>(pi / 2),
          weight: 50.0,
        ),
      ],
    ).animate(_flipController);
    _backScale = TweenSequence(
      <TweenSequenceItem<double>>[
        TweenSequenceItem<double>(
          tween: ConstantTween<double>(pi / 2),
          weight: 50.0,
        ),
        TweenSequenceItem<double>(
          tween: Tween(begin: -pi / 2, end: 0.0)
              .chain(CurveTween(curve: Curves.easeOut)),
          weight: 50.0,
        ),
      ],
    ).animate(_flipController);
    _flipController.addStatusListener((status) {
      if (status == AnimationStatus.completed ||
          status == AnimationStatus.dismissed) {
        if (onFlipDone != null) onFlipDone(isFront);
      }
    });
  }

  void initialiseAnimationAngle() {
    animation = Tween(begin: _angle, end: 0.0).animate(CurvedAnimation(
      parent: animationController,
      curve: Curves.bounceOut,
    ))
      ..addListener(() {
        setState(() {});
      });
  }

  _startDrag(DragStartDetails details) {
    setState(() {
      _start = details.globalPosition;
    });
  }

  _updateDrag(DragUpdateDetails details) {
    _end = details.globalPosition;
    setState(() {
      _angle = atan2(_end.dy - _start.dy, _end.dx - _start.dx) * (180 / pi);
      var _angle1 = atan2(0 - _start.dy, 0);
      var _angle2 = atan2(0 - _end.dy, _start.dx - _end.dx);
      _angle = 3 * (_angle1 - _angle2);

      _angle = -_angle;
      // if (_angle < 0) {
      //   _angle = 0;
      // }
      // _angle = 1;
    });
  }

  _endDrag(DragEndDetails details) async {
    try {
      setState(() {
        startAnim = true;
      });
      initialiseAnimationAngle();

      animationController.forward(from: 0.0).orCancel.then((value) {
        setState(() {
          startAnim = false;
          _angle = 0;
        });
      });
    } on TickerCanceled {}
  }

  _openCard() {
    if (onFlip != null) {
      onFlip();
    }
    if (isFront) {
      _flipController.forward(from: 0.0).orCancel.then((onValue) {
        setState(() {
          isFront = !isFront;
        });
      });
    } else {
      _flipController.reverse().orCancel.then((onValue) {
        setState(() {
          isFront = !isFront;
        });
      });
    }
  }

  _navigateCard() {
    //Navigate to the screen
  }

  @override
  Widget build(BuildContext context) {
    double _animationAngle;
    if (widget.moveListItem) {
      if (startAnim) {
        _animationAngle = animation.value;
      } else {
        _animationAngle = _angle;
      }
    }

    return Center(
      child: GestureDetector(
          //You can add your own gestures and their functionalites
          onHorizontalDragStart: widget.moveListItem ? _startDrag : null,
          onHorizontalDragUpdate: widget.moveListItem ? _updateDrag : null,
          onHorizontalDragEnd: widget.moveListItem ? _endDrag : null,
          onTap: _openCard,
          onDoubleTap: _navigateCard,
          child: Stack(
            fit: StackFit.passthrough,
            children: <Widget>[
              _buildContent(
                  isUpSide: true,
                  animationAngle: widget.moveListItem ? _animationAngle : 0),
              _buildContent(
                  isUpSide: false,
                  animationAngle: widget.moveListItem ? _animationAngle : 0)
            ],
          )),
    );
  }

  Widget _buildContent({@required bool isUpSide, double animationAngle}) {
    return IgnorePointer(
      ignoring: isUpSide ? !isFront : isFront,
      child: AnimationCard(
          animation: isUpSide ? _frontScale : _backScale,
          hangingObject: isUpSide ? widget.front : widget.back,
          height: widget._height,
          width: widget._width,
          animationAngle: animationAngle,
          isUpside: isUpSide),
    );
  }

  @override
  void dispose() {
    animationController.dispose();
    _flipController.dispose();
    super.dispose();
  }
}


lib/widgets/animationCard.dart
import 'package:flutter/material.dart';
import 'hangTransform.dart';

class AnimationCard extends StatelessWidget {
  final Widget hangingObject;
  final Animation<double> animation;
  final double height, width, animationAngle;
  final bool isUpside;
  AnimationCard(
      {this.animation,
      this.hangingObject,
      this.height,
      this.width,
      this.animationAngle,
      this.isUpside});
  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: animation,
      builder: (BuildContext context, Widget child) {
        final Matrix4 transform = new Matrix4.identity()
          ..setEntry(3, 2, 0.001)
          ..rotateY(animation.value);

        return HangTransform(
            transform, height, width, animationAngle, hangingObject, isUpside);
      },
    );
  }
}


lib/widgets/hangTransform.dart
import 'package:flutter/material.dart';

class HangTransform extends StatelessWidget {
  final Matrix4 transform;
  final double _height, _width, _animationAngle;
  final Widget _widget;
  final bool isUpside;
  HangTransform(this.transform, this._height, this._width, this._animationAngle,
      this._widget, this.isUpside);

  @override
  Widget build(BuildContext context) {
    return Transform(
        transform: transform,
        child: Container(
            height: _height,
            width: _width,
            child: Transform.rotate(
                angle: _animationAngle,
                child: Container(
                    decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(20.0),
                      boxShadow: <BoxShadow>[
                        BoxShadow(
                            color: Colors.black12,
                            blurRadius: 20.0,
                            offset: Offset(20.0, 20.0))
                      ],
                      gradient: LinearGradient(
                        colors: [Color(0xfffff9c4), Color(0xfffffde7)],
                        begin: FractionalOffset(1.0, 1.0),
                        end: FractionalOffset(1.7, 0.0),
                        stops: [0.0, 1.0],
                      ),
                    ),
                    child: _widget))));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
// import your package here
import 'package:hanginglist/hanginglist.dart';
import './model/hangingObj.dart';

void main() => runApp(MaterialApp(
      title: 'TASTYBURGER',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(primarySwatch: Colors.yellow),
      home: MyStateWidget(),
    ));

class MyStateWidget extends StatelessWidget {
  double screenWidth, screenHeight;

  @override
  Widget build(BuildContext context) {
    MediaQueryData mediaQueryData = MediaQuery.of(context);
    screenWidth = mediaQueryData.size.width;
    screenHeight = mediaQueryData.size.height;
    return Scaffold(
      appBar: AppBar(
        centerTitle: true,
        title: Text(
          'HANGINGLIST',
          style: TextStyle(
            fontWeight: FontWeight.bold,
            color: Colors.black,
          ),
          textAlign: TextAlign.center,
        ),
        actions: <Widget>[
          Icon(Icons.fastfood),
          Padding(
            padding: EdgeInsets.all(5),
          )
        ],
      ),
      // Define the HangingList Widget and do not forget to initialise all fields
      body: HangingList<HangingObject>(
        list: list,
        frontSide: frontSide,
        backSide: backSide,
        moveList: true,
        moveListItem: true,
        height: screenHeight / 2,
        width: screenWidth / 1.3,
      ),
    );
  }

// Create a generic function which returns the frontSide widget of the card
  Widget frontSide<HangingObject>(obj) {
    return Container(
        padding: EdgeInsets.all(10),
        child: Column(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: <Widget>[
            Align(
              alignment: Alignment.center,
              child: Text(
                obj.restaurantName,
                style: TextStyle(fontSize: 30, color: Color(0xffb71c1c)),
              ),
            ),
            SizedBox(
              height: 10,
            ),
            SizedBox(
                width: 200,
                child: RaisedButton(
                  color: Color(0xffffab40),
                  onPressed: () {},
                  child: Text(
                    'Add to your order',
                    style: TextStyle(color: Colors.white),
                  ),
                  shape: RoundedRectangleBorder(
                      borderRadius: new BorderRadius.circular(18.0),
                      side: BorderSide(color: Color(0xffff6d00))),
                ))
          ],
        ));
  }

// Create a generic function which returns the backSide widget of the card
  Widget backSide<HangingObject>(obj) {
    return Padding(
      padding: EdgeInsets.all(10),
      child: Column(
        children: <Widget>[
          Align(
            alignment: Alignment.center,
            child: Text(obj.restaurantName,
                style: TextStyle(
                    fontSize: 24,
                    fontWeight: FontWeight.w600,
                    color: Color(0xffb71c1c))),
          ),
          SizedBox(
            height: 10,
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Text('Dishes:',
                style: TextStyle(
                    fontSize: 20,
                    color: Colors.brown,
                    fontWeight: FontWeight.w600)),
          ),
          ListView.builder(
            shrinkWrap: true,
            itemCount: obj.dishes.length,
            itemBuilder: (BuildContext context, int i) {
              return Text(obj.dishes[i], style: TextStyle(fontSize: 16));
            },
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Text('Location:',
                style: TextStyle(
                    fontSize: 20,
                    color: Colors.brown,
                    fontWeight: FontWeight.w600)),
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Text(obj.location),
          ),
          SizedBox(
            height: 10,
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Text('Contact:',
                style: TextStyle(
                    fontSize: 20,
                    color: Colors.brown,
                    fontWeight: FontWeight.w600)),
          ),
          Align(
            alignment: Alignment.centerLeft,
            child: Text(obj.contact),
          ),
          SizedBox(
            height: 10,
          ),
          SizedBox(
              width: 200,
              child: RaisedButton(
                color: Color(0xffffab40),
                onPressed: () {},
                child: Text(
                  'ORDER NOW',
                  style: TextStyle(color: Colors.white),
                ),
                shape: RoundedRectangleBorder(
                    borderRadius: new BorderRadius.circular(18.0),
                    side: BorderSide(color: Color(0xffff6d00))),
              ))
        ],
      ),
    );
  }

// Create a list of objects which hold your desired data and also not forget to define your model
  List<HangingObject> list = [
    HangingObject(
        'Chicken',
        'chicken',
        'McDonald`s',
        ['Chicken McNuggets', 'Big Mac', 'Fries'],
        'Fusion Mall,2ndStage, Kormangala',
        '1860 210 0000, 080 3399 4444'),
    HangingObject(
        'Burger',
        'burger',
        'Taco Bell',
        ['Doritos Locos Tacos', 'Freezes', 'Nacho Fries', 'The Meximelt'],
        'Ashwini Complex, Indranagar',
        '1860 210 0000, 080 3399 4444'),
    HangingObject(
        'Pizza',
        'pizza',
        'KFC',
        [
          'Boneless Wings',
          'Extra Crispy Strips',
          'Extra Crispy Chicken',
          'Hot Wings'
        ],
        'BDA Complex HSR',
        '1860 210 0000, 080 3399 4444'),
    HangingObject(
        'Fries',
        'fries',
        'Burger King',
        ['Whopper Jr. Sandwich', 'HamBurger', 'Veggie Burger', 'TENDERGRILL'],
        'Bull Temple Road, Basavanagudi',
        '1860 210 0000, 080 3399 4444'),
    HangingObject(
        'Coke',
        'coke',
        'Subway',
        [
          'Rotisserie-Style Chicken',
          'Veggie Delite',
          'Meatball Marinara',
          'Meatball Marinara'
        ],
        'Fusion Mall,2ndStage, Kormangala',
        '1860 210 0000, 080 3399 4444'),
    HangingObject(
        'Sandwich',
        'sandwich',
        'StarBucks',
        ['Cake pops', 'Pumpkin bread, Greek yogurt, Bacon'],
        'Orion East, Banaswadi',
        '1860 210 0000, 080 3399 4444')
  ];
}
